### ISSUE TITLE
Large MatMuls use slow reference kernel on ARM

### ISSUE THREAD

**Issue by NishantPrabhuFujitsu** [openvinotoolkit/openvino.genai#438]
@Wovchena This is related to #406 but goes deeper, hence I decided to make this a new issue.

## TL;DR
- For `greedy_causal_lm` inference on `arm`, large matmuls (e.g. `1x2x4096:4096x4096` in query/key/value projection) fall back to using a slow reference matmul implementation (`ref_any_bf16`) 
- Smaller matmuls (e.g. `1x32x2x2:1x32x2x128` in dot-product attention) use a faster `gemm_acl_f16` kernel, which indicates that ACL kernels through oneDNN are available but not being used for point (1). 
- On `x86` most matmuls use `brgemm_avx512_bf16` kernel resulting in much faster inference. 

**Question:** Is there a heuristic in OpenVINO that causes large matmuls on `arm` to fall back to the reference implementation?

## Details
Following #327, I added code in `greedy_causal_lm.cpp` to save profiling information after inference. The table below shows the kernels used on `x86` vs `arm` and their runtime (in microseconds) from the first decoder block of `meta-llama/Llama-2-7b-hf` (HuggingFace, compressed to INT8_ASYM). 

| Operation           | Shape                | x86 kernel         | x86 time (mcs) | arm kernel  | arm time (mcs) |
|---------------------|----------------------|--------------------|----------------|-------------|----------------|
| q_proj              | 1x2x4096:4096x4096   | `brgemm_avx512_bf16` | 584            | `ref_any_f16` | 44714          |
| k_proj              | 1x2x4096:4096x4096   | `brgemm_avx512_bf16` | 699            | `ref_any_f16` | 39713          |
| v_proj              | 1x2x4096:4096x4096   | `brgemm_avx512_bf16` | 5775           | `ref_any_f16` | 37816          |
| o_proj              | 1x2x4096:4096x4096   | `brgemm_avx512_bf16` | 833            | `ref_any_f16` | 36469          |
| mlp.gate_proj       | 1x2x4096:4096x11008  | `brgemm_avx512_bf16` | 1023           | `ref_any_f16` | 97537          |
| mlp.up_proj         | 1x2x4096:4096x11008  | `brgemm_avx512_bf16` | 965            | `ref_any_f16` | 101677         |
| mlp.down_proj       | 1x2x11008:11008x4096 | `brgemm_avx512_bf16` | 953            | `ref_any_f16` | 98749          |
| classifier (logits) | 1x2x4096:4096x32000  | `brgemm_avx512_bf16` | 1813           | `ref_any_f16` | 284708         |

For small matmuls however, I noticed `gemm_acl_f16` being called on `arm`.

| Operation               | Shape                 | arm kernel   | arm time (mcs) |
|-------------------------|-----------------------|--------------|----------------|
| A = q_proj * k_proj ^ T | 1x32x2x128:1x32x2x128 | `gemm_acl_f16` | 226            |
| A * v_proj              | 1x32x2x2:1x32x2x128   | `gemm_acl_f16` | 69             |

This indicates that ACL kernels are available during inference but not being used **only for large matmuls**. To check that this diversion is not coming from oneDNN, I compiled oneDNN with ACL with following version combinations and benchmarked the above matmul sizes with `benchdnn`. Results are shown below, which indicate that `gemm:acl` is used on `arm` for all of the above sizes when the two compile correctly.

| oneDNN source          | oneDNN version | ACL version | Kernel used on arm          | Comments                                                |
|------------------------|----------------|-------------|-----------------------------|---------------------------------------------------------|
| oneapi-src/oneDNN      | 3.3.3          | 24.02.1     | Build fails; error from ACL | References to ACL v24.02.1 were found in openvino docs. |
| oneapi-src/oneDNN      | 3.3.3          | 23.11       | Build fails; error from ACL | N/A                                                     |
| oneapi-src/oneDNN      | 3.3.3          | 23.08       | `gemm:acl`                    | Expected kernel is used.                                |
| openvinotoolkit/oneDNN | 3.3.3          | 24.02.1     | `gemm:jit_f32`                | ACL compilation flag ignored by cmake                   |
| openvinotoolkit/oneDNN | 3.3.3          | 23.11       | `gemm:jit_f32`                | ACL compilation flag ignored by cmake                   |
| openvinotoolkit/oneDNN | 3.3.3          | 23.08       | `gemm:jit_f32`                | ACL compilation flag ignored by cmake                   | 

I think ACL kernels should ideally be used for any matmul on `arm` since they are available and being used for some operations. Since oneDNN is not the one causing a reference kernel to be used, I am led to believe that a heuristic within OpenVINO is causing this fallback. I'm hoping to know whether that is indeed the case, and (if yes) where this heuristic has been implemented.

**Comment by alvoron** [comment]
Hi @NishantPrabhuFujitsu 
Thank you for submitting the issue, I'm reviewing it.
What ARM did you use to reproduce the issue? Is it Apple M1/M2/M3 or something else?

**Comment by NishantPrabhuFujitsu** [comment]
@alvoron Thanks for taking it up. The ARM CPU I'm using are on Graviton3 instances on AWS.

**Comment by NishantPrabhuFujitsu** [comment]
Hi @alvoron. Have you gotten any insights on this yet?

**Comment by alvoron** [comment]
@NishantPrabhuFujitsu yes, I have some ideas what goes wrong.
Will you be able to check a custom OpenVINO build if I provide you OpenVINO wheel package?

**Comment by alvoron** [comment]
@NishantPrabhuFujitsu 
Since you're using C++ API you need C++ distribution of OpenVINO.
I've prepared a branch with needed fix - https://github.com/alvoron/openvino/tree/alvoron_acl_master
Could you please build OpenVINO on this branch?
OpenVINO build procedure (you don't need additional build options, just the first 4 steps): https://github.com/openvinotoolkit/openvino/blob/master/docs/dev/build_linux.md
And then 1 installation command: https://github.com/openvinotoolkit/openvino/blob/master/docs/dev/installing.md

**Comment by NishantPrabhuFujitsu** [comment]
Hi @alvoron.

The fix worked! Thank you so much for your support. The execution times for MatMul (and FullyConnected) layers are much better now. I found out from the conversation [here](https://github.com/openvinotoolkit/openvino/issues/7273#issuecomment-2124714531) that this was an ACL bug. If I may ask, what was the bug, and what was done to fix it?

Thanks again for helping out.

**Comment by NishantPrabhuFujitsu** [comment]
Marking as closed since the issue has been fixed.

**Comment by alvoron** [comment]
@NishantPrabhuFujitsu 
The issue was related to ACL multi-isa feature: ACL could be built using one ISA (e.g. without FP16 support), however this build could be run on a machine with another ISA (with FP16 support) and FP16 kernels should be available. The issue is some FP16 kernels are not compiled on ISA without FP16.
It's already fixed in ACL developer repository, however, there is no ACL release with this fix. My branch uses ACL Release Candidate that contains needed changes. 

Please feel free to contact me if you face any issues or have any requests/suggestions on ARM functionality/performance.

**Comment by NishantPrabhuFujitsu** [comment]
@alvoron Thanks for letting me know. I'll be sure to get in touch if I need more help with running this on ARM.

Just one more request: would it be possible for you to send a `.whl` file for python 3.11.0 with your changes? I need to perform some benchmarking through the Python API as well. It's also okay if you tell me how I can build it locally.

**Comment by alvoron** [comment]
@NishantPrabhuFujitsu 
You can build wheel locally. You should follow the same build procedure you used before to build C++ package.
The only difference is to add `-DENABLE_PYTHON=ON -DENABLE_WHEEL=ON` to cmake configure command.
You find whl file in `<INSTALL_DIR>/tools` directory.
Please let me know if you face any issues, I'll help.

**Comment by alvoron** [comment]
@NishantPrabhuFujitsu 
1. Definitely we don't expect to see significant performance results between Python and C++. I'd suggest you to create a dedicated issue for that, we need to analyse this case.
2. OpenVINO on ARM supports fp16 and fp32 natively. bf16 is not supported, I need to review why bf16 is mentioned in the response. Is there any reason you'd prefer bf16 rather than fp16? i8 is not supported as well, however we're already working on i8 support.
3. We expect all kernels are taken from ACL or oneDNN. I'll double check with colleagues what component brings OpenBLAS dependency. This method consumes only 0.02% - it seems to me it's not kernel execution.

**Comment by NishantPrabhuFujitsu** [comment]
@alvoron Hope you had a good weekend. 

1. I discovered something regarding the speed difference between Python and C++ inference (code linked below point to scripts in my fork of your repo).

| Language | Implementation overview                                                                          | Link to code/script | Throughput (tokens/sec) |
|----------|--------------------------------------------------------------------------------------------------|---------------------|-------------------------|
| C++      | Ported and built `greedy_causal_lm.cpp` with necessary changes from openvino.genai into `samples/` | [link](https://github.com/NishantPrabhuFujitsu/openvino/tree/llm_inference/samples/cpp/greedy_causal_lm)                | 6.47                    |
| Python   | Using `OVModelForCausalLM` from HuggingFace `optimum.intel`                    | [link](https://github.com/NishantPrabhuFujitsu/openvino/blob/llm_inference/samples/python/greedy_causal_lm/with_optimum.py)                | 0.07                    |
| Python   | Using `infer_request` API with asynchronous inference from `openvino` package directly                                                  | [link](https://github.com/NishantPrabhuFujitsu/openvino/blob/llm_inference/samples/python/greedy_causal_lm/with_ov_core.py)                | 7.70                    |

It appears using `infer_request.start_async()` for inference delivers the best performance on ARM so far (comparable to that on x86). Perhaps `optimum` uses synchronous/non-threaded inference making it slow? Let me know if you have any ideas. Secondly, python inference is now (somewhat) faster than C++; any ideas on that? I ran the experiment many times and the third implementation mentioned above was the fastest every time. In any case, I don't think I should open a new issue for this since the problem doesn't seem to have to do with `openvino` now.

2. I've experienced better model output quality (accuracy?) with bf16 precision compared to fp16 in the past, hence it has higher preference vs fp16. Also, great to know that int8 support is on the way.
3. Please find the entire `perf` report here: [t5_perf.txt](https://github.com/openvinotoolkit/openvino.genai/files/15454188/t5_perf.txt)

**Comment by alvoron** [comment]
@eaidova could you please comment `optimum` behaviour, that @NishantPrabhuFujitsu highlighted above?

**Comment by dmitry-gorokhov** [comment]
@NishantPrabhuFujitsu Let me try to help with your questions:
1.1 I would not expect any perf issues with optimum.intel which should have more or less the same OV calls underneath as in your third example. Have you checked your python env? Maybe it is still old OV version is used (w/o the fix from @alvoron)?
1.2 I would not expect C++ API to be slower than Python. I guess we can look on that.
As a side note I would recommend to take a look on our LLM bench which helps to evaluate performance in different scenarious (greedy/beam_search, different prompt or output size, separate timing for prompt processing and generate stages etc): https://github.com/openvinotoolkit/openvino.genai/tree/master/llm_bench/python
Additionally we are finalizing C++ Generate API for genai, might be interesting for you: https://github.com/openvinotoolkit/openvino.genai/pull/334

2. Based on our experience some models shows better accuracy with fp16 and some with bf16, so it depends. The tricky thing is performace. AFAIK ACL has limited BF16 support (as well as HW itself), so bf16 inference might not be as performant as fp16. Anyway this is smt we can consider as an option for the future development.

3. Based on the log sgemm is called as part of torch inference, not OpenVINO. Pytorch infenrece is used for tokenizer model and as part of beam search algo inside generate call.

**Comment by NishantPrabhuFujitsu** [comment]
@dmitry-gorokhov @alvoron Thanks for your response.

1. Regarding performance related points:
    - The poor throughput from `optimum` is most likely due to the right openvino package not being installed, like you pointed out. I ran several experiments today but I can't derive any conclusions right now due to large variations in performance I'm observing for the other two experiments (compared to yesterday). This variation is due to issues with the instances I'm using for my experiments, not with OpenVINO. I'll try and resolve them at my end and come back to you with some reliable data soon.
    -  Between C++ and Python API, I noticed that the python script (third experiment in my previous message) was slightly faster than C++ even on x86. For example, in one experiment the C++ script yielded 13.40 tokens/sec while the Python script yielded 14.02 tokens/sec. Regardless, I will take these readings again once I resolve the issue I mentioned above.
    - I tried out `llm_bench` on ARM with Llama2, and got roughly the same performance numbers as I did from my script. The output of `llm_bench` is much more informative however, so I'll probably continue using that going forward 😄 .
    - #334 -> very useful; eagerly awaited.

2. I discussed with some of my colleagues, and they found it interesting that FP16 and BF16 had comparable accuracy as per your findings. We've almost always had BF16 give better accuracy, across many architectures (not limited to LLMs) and tasks. However, I can understand that there could be situations where this is not the case. It's fine to take this up at a later point; not very urgent.
3. Thanks for clarifying.

**Comment by NishantPrabhuFujitsu** [comment]
@alvoron @dmitry-gorokhov I hope you're doing well.

1. I spent some time correcting my benchmarking procedure, and have the following statistics summarized from ~10 runs. Below, "Precision" refers to inference precision specified using the inference precision hint.

<img width="602" alt="benchmark" src="https://github.com/openvinotoolkit/openvino.genai/assets/168545601/2c2b1e23-6dfa-4961-bd92-72df4eca33b3">

Based on the above, I can say the following.
- The python implementation is about as fast as the C++ one, but the C++ is slightly faster on an average as expected. So my point about Python being faster than C++ mentioned earlier stands corrected.
- About `optimum`: I am fairly confident there's an issue with inference using `optimum`. In several experiments where I ensured that the right openvino whl file is installed (on ARM), I get much poorer performance using `optimum` than with C++ or python using `openvino` directly (see benchmark below). Perhaps this could be investigated further?

|                         | arm (Graviton3E) | x86 (Sapphire Rapids) |
|:-----------------------:|:----------------:|:---------------------:|
| Throughput (tokens/sec) |      1.8423      |         8.3086        |

2. When you mentioned earlier that int8 inference is being worked on, was that only for arm? I noticed that on x86 (Sapphire Rapids) not specifying any inference precision hint or using inference precision FP32 with `DNNL_DEFAULT_FPMATH_MODE=BF16` generates oneDNN logs that show weight tags with "u8" precision:
```
onednn_verbose,primitive,exec,cpu,inner_product,brgemm:avx512_core,forward_inference,src_f32::blocked:ab::f0 wei_u8:a:blocked:AB16b64a::f0 bia_undef::undef::: dst_f32::blocked:ab::f0,attr-scratchpad:user attr-fpmath:bf16 attr-scales:wei:1 attr-zero-points:wei:1 ,,mb1ic4096oc4096,0.154053
onednn_verbose,primitive,exec,cpu,inner_product,brgemm:avx512_core,forward_inference,src_f32::blocked:ab::f0 wei_u8:a:blocked:AB16b64a::f0 bia_undef::undef::: dst_f32::blocked:ab::f0,attr-scratchpad:user attr-fpmath:bf16 attr-scales:wei:1 attr-zero-points:wei:1 ,,mb1ic4096oc11008,0.419189
onednn_verbose,primitive,exec,cpu,inner_product,brgemm:avx512_core,forward_inference,src_f32::blocked:ab::f0 wei_u8:a:blocked:AB16b64a::f0 bia_undef::undef::: dst_f32::blocked:ab::f0,attr-scratchpad:user attr-fpmath:bf16 attr-scales:wei:1 attr-zero-points:wei:1 attr-post-ops:eltwise_swish:1 ,,mb1ic4096oc11008,0.422852
onednn_verbose,primitive,exec,cpu,inner_product,brgemm:avx512_core,forward_inference,src_f32::blocked:ab::f0 wei_u8:a:blocked:AB16b64a::f0 bia_undef::undef::: dst_f32::blocked:ab::f0,attr-scratchpad:user attr-fpmath:bf16 attr-scales:wei:1 attr-zero-points:wei:1 ,,mb1ic11008oc4096,0.39209
onednn_verbose,primitive,exec,cpu,inner_product,brgemm:avx512_core,forward_inference,src_f32::blocked:ab::f0 wei_u8:a:blocked:AB16b64a::f0 bia_undef::undef::: dst_f32::blocked:ab::f0,attr-scratchpad:user attr-fpmath:bf16 attr-scales:wei:1 attr-zero-points:wei:1 ,,mb1ic4096oc4096,0.155029
onednn_verbose,primitive,exec,cpu,inner_product,brgemm:avx512_core,forward_inference,src_f32::blocked:ab::f0 wei_u8:a:blocked:AB16b64a::f0 bia_undef::undef::: dst_f32::blocked:ab::f0,attr-scratchpad:user attr-fpmath:bf16 attr-scales:wei:1 attr-zero-points:wei:1 ,,mb1ic4096oc4096,0.147949
onednn_verbose,primitive,exec,cpu,inner_product,brgemm:avx512_core,forward_inference,src_f32::blocked:ab::f0 wei_u8:a:blocked:AB16b64a::f0 bia_undef::undef::: dst_f32::blocked:ab::f0,attr-scratchpad:user attr-fpmath:bf16 attr-scales:wei:1 attr-zero-points:wei:1 ,,mb1ic4096oc4096,0.152832
```
Just wanted to check whether the above means that operations are being performed at int8 precision. 

3. Do you have any idea how one can set the number of threads to use for inference for `optimum`? I looked through the documentation for `optimum.intel` and the code for quite some time but couldn't get this to work. I also have the same question about the correct way of setting inference precision.

## How inference precision and number of threads were set
I'm including this to show how I set inference precision hints and thread count (in appropriate sections of the code). If this is not the right way, please point me to the correct method.

**C++**
```cpp
ov::CompiledModel compiled_model = core.compile_model(
    std::string{argv[1]} + "/openvino_model.xml", 
    "CPU",
    ov::inference_num_threads(nthr), // read and converted to uint from argv
    ov::hint::inference_precision(ov::element::f32) // or f16
);
```
**Python**
```python
import openvino as ov
import openvino.properties as props
import openvino.properties.hint as hints

core = ov.Core()
config = {
    props.inference_num_threads: args.num_threads, # read from CLI args
    hints.inference_precision: "f32", # or f16
}

model = core.read_model(f"{MODEL_DIR}/openvino_model.xml")
compiled_model = core.compile_model(model=model, device_name="CPU", config=config)
lm = compiled_model.create_infer_request()
```
